File: bob/learn/boosting/cpp/LossFunction.cpp
#include <bob.learn.boosting/LossFunction.h>
#include <math.h>

void bob::learn::boosting::LossFunction::lossSum(const blitz::Array<double,1>& alpha, const blitz::Array<double,2>& targets, const blitz::Array<double,2>& previous_scores, const blitz::Array<double,2>& current_scores, blitz::Array<double,1>& loss_sum) const{
  // compute the scores and loss for the current alpha
  scores.resize(targets.shape());
  // TODO: is there any faster way for this?
  for (int i = scores.extent(0); i--;){
    for (int j = scores.extent(1); j--;){
      scores(i,j) = previous_scores(i,j) + alpha(j) * current_scores(i,j);
    }
  }
  errors.resize(targets.extent(0), 1);
  loss(targets, scores, errors);

  // compute the sum of the loss
  blitz::firstIndex i;
  blitz::secondIndex j;
  loss_sum = blitz::sum(errors(j,i), j);
}


void bob::learn::boosting::LossFunction::gradientSum(const blitz::Array<double,1>& alpha, const blitz::Array<double,2>& targets, const blitz::Array<double,2>& previous_scores, const blitz::Array<double,2>& current_scores, blitz::Array<double,1>& gradient_sum) const{
  // compute the scores and gradient for the current alpha
  scores.resize(targets.shape());
  // TODO: is there any faster way for this?
  for (int i = scores.extent(0); i--;){
    for (int j = scores.extent(1); j--;){
      scores(i,j) = previous_scores(i,j) + alpha(j) * current_scores(i,j);
    }
  }
//  scores = previous_scores + alpha * current_scores;

  gradients.resize(targets.shape());
  lossGradient(targets, scores, gradients);

  // take the sum of the loss gradient values
  const blitz::Array<double, 2> grad(gradients * current_scores);
  blitz::firstIndex i;
  blitz::secondIndex j;
  gradient_sum = blitz::sum(grad(j,i), j);
}

File: bob/learn/boosting/LossFunction.py
import numpy


class LossFunction(object):
    """This is a base class for all loss functions implemented in pure python.
    It is simply a python re-implementation of the :py:class:`bob.learn.boosting.LossFunction` class.

    This class provides the interface for the L-BFGS optimizer.
    Please overwrite the loss() and loss_gradient() function (see below) in derived loss classes.
    """

    def __init__(self):
        pass

    def loss(self, targets, scores):
        """This function is to compute the loss for the given targets and scores.

        Keyword parameters:

          targets (float <#samples, #outputs>): The target values that should be reached.

          scores (float <#samples, #outputs>): The scores provided by the classifier.

        Returns
          (float <#samples, #outputs>) or (float <#samples, 1>): The loss based on the given scores and targets.
          Depending on the intended task, one of the two output variants should be chosen.
          For classification tasks, please use the former way (#samples, #outputs), while for regression tasks, use the latter (#samples, 1).
        """
        raise NotImplementedError("This is a pure abstract function. Please implement that in your derived class.")

    def loss_gradient(self, targets, scores):
        """This function is to compute the gradient of the loss for the given targets and scores.

        Keyword parameters:

          targets (float <#samples, #outputs>): The target values that should be reached.

          scores (float <#samples, #outputs>): The scores provided by the classifier.

        Returns
          loss (float <#samples, #outputs>): The gradient of the loss based on the given scores and targets.
        """
        raise NotImplementedError("This is a pure abstract function. Please implement that in your derived class.")

    def loss_sum(self, alpha, targets, previous_scores, current_scores):
        """The function computes the sum of the loss which is used to find the optimized values of alpha (x).

        The functions computes sum of loss values which is required during the line search step for the optimization of the alpha.
        This function is given as the input for the L-BFGS optimization function.

        Keyword parameters:

          alpha (float): The current value of the alpha.

          targets (float <#samples, #outputs>): The targets for the samples

          previous_scores (float <#samples, #outputs>): The cumulative prediction scores of the samples until the previous round of the boosting.

          current_scores (float <#samples, #outputs>): The prediction scores of the samples for the current round of the boosting.

        Returns

          (float <#outputs>) The sum of the loss values for the current value of the alpha
        """

        # compute the scores and loss for the current alpha
        scores = previous_scores + alpha * current_scores
        losses = self.loss(targets, scores)

        # compute the sum of the loss
        return numpy.sum(losses, 0)

    def loss_gradient_sum(self, alpha, targets, previous_scores, current_scores):
        """The function computes the gradient as the sum of the derivatives per sample which is used to find the optimized values of alpha.

        The functions computes sum of loss values which is required during the line search step for the optimization of the alpha.
        This function is given as the input for the L-BFGS optimization function.

        Keyword parameters:

          alpha (float): The current value of the alpha.

          targets (float <#samples, #outputs>): The targets for the samples

          previous_scores (float <#samples, #outputs>): The cumulative prediction scores of the samples until the previous round of the boosting.

          current_scores (float <#samples, #outputs>): The prediction scores of the samples for the current round of the boosting.

        Returns
          (float <#outputs>) The sum of the loss gradient for the current value of the alpha.
        """

        # compute the loss gradient for the updated score
        scores = previous_scores + alpha * current_scores
        loss_gradients = self.loss_gradient(targets, scores)

        # take the sum of the loss gradient values
        return numpy.sum(loss_gradients * current_scores, 0)
